thorlcr/mfilemanager/thmfilemanager.cpp

```diff
@@ -145,7 +146,57 @@ class CFileManager : public CSimpleInterface, implements IThorFileManager
         return max;
     }
 
-    void remove(IDistributedFile &file, IUserDescriptor &userDesc)
+    template<typename T>
+    T blockReportFunc(CJobBase &job, std::function<T(unsigned timeout)> func, unsigned timeout, const char *msg)
+    {
+        // every minute timeout, and set workunit state to blocked for NN minutes
+        WUState preBlockedState = WUStateUnknown; // set to current state if times out and blocks
+        CTimeMon tm(timeout);
+        if (timeout >= 60000)
+            timeout = 60000;
+        while (true)
+        {
+            try
+            {
+                T ret = func(timeout);
+                if (WUStateUnknown != preBlockedState)
+                {
+                    Owned<IWorkUnit> workunit = &job.queryWorkUnit().lock();
+                    WUState currentState = workunit->getState();
+                    if (WUStateBlocked != currentState) // could happen if user aborted whilst blocked. Bail out.
+                        throw MakeStringException(0, "WorkUnit state has changed to '%s' whilst blocked trying to '%s'", getWorkunitStateStr(currentState), msg);
+                    workunit->setState(preBlockedState); // i.e. restore original state
+                }
+                return ret;
+            }
+            catch (ISDSException *e)
+            {
+                if ((SDSExcpt_LockTimeout != e->errorCode()) || tm.timedout())
+                    throw;
+                e->Release();
+            }
+
+            // NB: will reach here only if timeout occurred.
+            Owned<IWorkUnit> workunit = &job.queryWorkUnit().lock();
+            WUState currentState = workunit->getState();
+            if (WUStateUnknown == preBlockedState) // i.e. 1st time
+            {
+                if (WUStateUnknown == currentState) // should never happen, but guard against it JIC
+                    throw MakeStringException(0, "WorkUnit in unknown state");
+                preBlockedState = currentState;
+                workunit->setState(WUStateBlocked);
+            }
+            else // NB: implies >= 2nd time around
+            {
+                if (WUStateBlocked != currentState) // could happen if user aborted whilst blocked. Bail out.
+                    throw MakeStringException(0, "WorkUnit state has changed to '%s' whilst blocked trying to '%s'", getWorkunitStateStr(currentState), msg);
+            }
+            VStringBuffer blockedMsg("Blocked for %u minutes trying to %s", tm.elapsed()/60000, msg);
```

Comment:
Would timeout ever be less than 60000 ?

Reply:
It would never get here if < 1 minute.
The timeout exception would be rethrown at line 175 because it would have already timed out.